### PR TITLE
[REVIEW]Cupy fallback for __array_function__  and __array_ufunc__ for cudf.Series 

### PR DESCRIPTION
**Pull request by VibhuJawa**
This pr adds a cupy fallback for   `__array_function__`  and `__array_ufunc__ `. It closes https://github.com/rapidsai/cudf/issues/2718 .  

The idea is that  if a user calls a numpy `__array_ufunc__ ` or `__array_function__` on `cudf.Series` and we fallback on cupy for implementation  if we don't have it in `cudf.Series`. 

Tests to add: 
- [x] Check if null present in series where function is dispatched 
- [x] Check for `np` function on`cudf.Series` and `cupy array` . 
- [x] Test for `scalar` output 
- [x] Test for `list` inputs

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

File: python/cudf/cudf/utils/utils.py [hunk 1]
```diff
@@ -464,3 +466,64 @@ def search_range(start, stop, x, step=1, side="left"):
...
+        cupy_output = cupy_func(*cupy_compatible_args, **kwargs)
+        return cast_to_appropriate_cudf_type(cupy_output)
+    else:
+        return NotImplemented
+
+
+def cast_to_appropriate_cudf_type(val):
+    # TODO Handle scalar
+    if val.ndim == 0:
+        return cudf.Scalar(val).value
```
Comment (VibhuJawa) [inline python/cudf/cudf/utils/utils.py:497]: Unsure what should I output when the output of the NumPy function is a scalar but dispatched cupy function creates a zero-dimensional array.

**Comment by VibhuJawa** [comment]
rerun tests

**Comment by VibhuJawa** [comment]
rerun tests.

File: python/cudf/cudf/utils/utils.py (hunk 1 above)
Comment (galipremsagar) [reply python/cudf/cudf/utils/utils.py:497]: I think returning `.value` seems to be a valid approach for interop with other libraries. @brandon-b-miller do you think we must return only a `cudf.Scalar` instead?

**Review by galipremsagar — APPROVED** [review]
Changes look good to me.

**Comment by galipremsagar** [comment]
rerun tests

File: python/cudf/cudf/utils/utils.py (hunk 1 above)
Comment (brandon-b-miller) [reply python/cudf/cudf/utils/utils.py:497]: calling `.value` here will give you back a host side numpy scalar. Is the intention to create a scalar from a 0-D cupy array without incurring a host transfer?

Comment (VibhuJawa) [reply python/cudf/cudf/utils/utils.py:497]: The goal here was to emulate the `cuDF.Series` functions that return scalar and just follow that behavior. We currently return `numpy scalars` ( See the below example).   Happy to change it to what ever you guys recommend. 

```python
>>> import cudf
>>> o = cudf.Series([1,2,3]).sum()
>>> type(o)
<class 'numpy.int64'>
```

Comment (brandon-b-miller) [reply python/cudf/cudf/utils/utils.py:497]: Ah, I see. For at least one release most likely, we'll keep returning numpy scalars for reduction ops like `sum`. This code should work fine, under the hood constructing a `cudf.Scalar` will call `val.item()` in the 0-D cupy array case and return that on the host.

Comment (VibhuJawa) [reply python/cudf/cudf/utils/utils.py:497]: Gotcha, Thanks a lot for the inputs @brandon-b-miller ,  @galipremsagar . I am leaving the code as-is. We can change this to return `cudf scalar` once we change behavior for all reduction ops. Resolving this.

File: python/cudf/cudf/core/series.py [hunk 2]
```diff
@@ -783,39 +785,25 @@ def __len__(self):
         return len(self._column)
 
     def __array_ufunc__(self, ufunc, method, *inputs, **kwargs):
-        if method == "__call__" and hasattr(cudf, ufunc.__name__):
-            func = getattr(cudf, ufunc.__name__)
-            return func(*inputs)
+        if method == "__call__":
+            return get_appropriate_dispatched_func(
+                cudf, cupy, ufunc, inputs, kwargs
```
Comment (VibhuJawa) [inline python/cudf/cudf/core/series.py]: FYI @galipremsagar I had to add both `cuDF`, `cuDF.series` as fallback options. 

I had earlier incorrectly switched this to`cudf.Series` from `cudf` and that led to some failures with tests like subtraction b/w `numpy` DateTime scalar and `cudf.Series` .   

Now we will first search `cudf` function space followed by `cudf.Series` function space, finally falling back to `cupy`.  Please let me know what you think. 

Previously we only looked at `cudf` name space for `ufunc` and `cudf.Series` for `array function`. IMO this should be a better solution.

File: python/cudf/cudf/core/series.py [hunk 3]
```diff
@@ -783,39 +785,31 @@ def __len__(self):
         return len(self._column)
 
     def __array_ufunc__(self, ufunc, method, *inputs, **kwargs):
-        if method == "__call__" and hasattr(cudf, ufunc.__name__):
-            func = getattr(cudf, ufunc.__name__)
-            return func(*inputs)
+        if method == "__call__":
+            return get_appropriate_dispatched_func(
+                cudf, cudf.Series, cupy, ufunc, inputs, kwargs
```
Comment (VibhuJawa) [inline python/cudf/cudf/core/series.py:790]: FYI @galipremsagar I had to add both cuDF, cuDF.series as fallback options.

I had earlier incorrectly switched this tocudf.Series from cudf and that led to some failures with tests like subtraction b/w numpy DateTime scalar and cudf.Series .

Now we will first search cudf function space followed by cudf.Series function space, finally falling back to cupy. Please let me know what you think.

Previously we only looked at cudf name space for ufunc and cudf.Series for array function. IMO this should be a better solution.

**Review by galipremsagar — APPROVED** [review]

**Pull request merged**